File: Java-Core/src/main/java/presto/statement/StatementList.java
package presto.statement;

import java.util.LinkedList;

import presto.error.NullReferenceError;
import presto.error.PrestoError;
import presto.error.SyntaxError;
import presto.java.JavaNativeCall;
import presto.parser.Dialect;
import presto.runtime.Context;
import presto.type.IType;
import presto.type.TypeMap;
import presto.type.VoidType;
import presto.utils.CodeWriter;
import presto.value.IValue;


public class StatementList extends LinkedList<IStatement> {

	private static final long serialVersionUID = 1L;

	public StatementList() {
	}

	public StatementList(IStatement stmt) {
		this.add(stmt);
	}
	
	@Override
	public boolean add(IStatement stmt) {
		if(stmt!=null)
			return super.add(stmt);
		else
			return false;
	}
	
	public IType check(Context context) throws SyntaxError {
		TypeMap types = new TypeMap();
		for(IStatement statement : this) {
			IType type = statement.check(context);
			if(type!=VoidType.instance())
				types.put(type.getName(), type);
		}
		return types.inferType(context);
	}

	public IType checkNative(Context context, IType returnType) throws SyntaxError {
		TypeMap types = new TypeMap();
		for(IStatement statement : this) {
			if(!(statement instanceof JavaNativeCall))
				continue;
			IType type = ((JavaNativeCall)statement).checkNative(context, returnType);
			if(type!=VoidType.instance())
				types.put(type.getName(), type);
		}
		return types.inferType(context);
	}

	public IValue interpret(Context context) throws PrestoError {
		return doInterpret(context);
	}

	private IValue doInterpret(Context context) throws PrestoError {
		for(IStatement statement : this) {
			context.enterStatement(statement);
			try {
				IValue result = statement.interpret(context);
				if(result!=null)
					return result;
			} finally {
				context.leaveStatement(statement);
			}
		}
		return null;
	}
	
	public IValue interpretNative(Context context, IType returnType) throws PrestoError {
		try {
			return doInterpretNative(context, returnType);
		} catch(NullPointerException e) {
			e.printStackTrace();
			throw new NullReferenceError();
		}
	}
	
	private IValue doInterpretNative(Context context, IType returnType) throws PrestoError {
		for(IStatement statement : this) {
			if(!(statement instanceof JavaNativeCall))
				continue;
			context.enterStatement(statement);
			try {
				IValue result = ((JavaNativeCall)statement).interpretNative(context, returnType);
				if(result!=null)
					return result;
			} finally {
				context.leaveStatement(statement);
			}
		}
		return null;
	}

	public void toDialect(CodeWriter writer) {
		for(IStatement statement : this) {
			statement.toDialect(writer);
			if(statement instanceof SimpleStatement) {
				if(writer.getDialect()==Dialect.O)
					writer.append(';');
				writer.newLine();
			}
		}
	}


	
}

File: Java-CodeServer/src/main/resources/js/prompto/statement/StatementList.js
var ObjectList = require("../utils/ObjectList").ObjectList;
var TypeMap = require("../type/TypeMap").TypeMap;
var VoidType = require("../type/VoidType").VoidType;
var Dialect = require("../parser/Dialect").Dialect;
var NullReferenceError = require("../error/NullReferenceError").NullReferenceError;
var SimpleStatement = require("./SimpleStatement").SimpleStatement;
var JavaScriptNativeCall = require("../javascript/JavaScriptNativeCall").JavaScriptNativeCall;

function StatementList(statement) {
	ObjectList.call(this);
	this.add(statement);
	return this;
}

StatementList.prototype = Object.create(ObjectList.prototype);
StatementList.prototype.constructor = StatementList;

StatementList.prototype.check = function(context, nativeOnly) {
	nativeOnly = nativeOnly || false;
	var types = new TypeMap();
	for(var i=0;i<this.length;i++) {
		var stmt = this[i];
		if(nativeOnly && !(stmt instanceof JavaScriptNativeCall)) {
			continue;
		}
		var type = stmt.check(context);
		if(type!==VoidType.instance) {
			types[type.name] = type;
		}
	}
	return types.inferType(context);
};

StatementList.prototype.interpret = function(context) {
	try {
		return this.doInterpret(context);
	} catch(e) {
		if(e instanceof ReferenceError) {
			throw e; // new NullReferenceError();
		} else {
			throw e;
		}
	}
};

StatementList.prototype.interpretNative = function(context, returnType) {
    try {
        return this.doInterpretNative(context, returnType);
    } catch(e) {
        if(e instanceof ReferenceError) {
            throw e; // new NullReferenceError();
        } else {
            throw e;
        }
    }
};

StatementList.prototype.doInterpret = function(context) {
	for(var i=0;i<this.length;i++) {
    	var stmt = this[i];
		context.enterStatement(stmt);
		try {
			var result = stmt.interpret(context);
			if(result!=null)
				return result;
		} finally {
			context.leaveStatement(stmt);
		}
	}
	return null;
};

StatementList.prototype.doInterpretNative = function(context, returnType) {
    for(var i=0;i<this.length;i++) {
        var stmt = this[i];
        if(!(stmt instanceof JavaScriptNativeCall))
            continue;
        context.enterStatement(stmt);
        try {
            var result = stmt.interpret(context, returnType);
            if(result!=null)
                return result;
        } finally {
            context.leaveStatement(stmt);
        }
    }
    return null;
};

StatementList.prototype.toDialect = function(writer) {
    for(var i=0;i<this.length;i++) {
        var stmt = this[i];
        stmt.toDialect(writer);
        if(stmt instanceof SimpleStatement) {
            if(writer.dialect==Dialect.O)
                writer.append(';');
            writer.newLine();
        }
    }
};

exports.StatementList = StatementList;
